src/core/functions/aggregate/llm_min_or_max.cpp

```diff
@@ -0,0 +1,210 @@
+#include <flockmtl/core/functions/prompt_builder.hpp>
+#include <templates/llm_min_or_max_prompt_template.hpp>
+#include "flockmtl/core/module.hpp"
+#include "flockmtl/core/model_manager/model_manager.hpp"
+#include "flockmtl/core/model_manager/tiktoken.hpp"
+
+namespace flockmtl {
+namespace core {
+
+void LlmMinOrMaxState::Initialize() {
+}
+
+void LlmMinOrMaxState::Update(const nlohmann::json &input) {
+    value.push_back(input);
+}
+
+void LlmMinOrMaxState::Combine(const LlmMinOrMaxState &source) {
+    value = std::move(source.value);
+}
+
+LlmMinOrMax::LlmMinOrMax(std::string &model, int model_context_size, std::string &user_prompt,
+                         std::string &llm_min_or_max_template)
+    : model(model), model_context_size(model_context_size), user_prompt(user_prompt),
+      llm_min_or_max_template(llm_min_or_max_template) {
+
+    int fixed_tokens = calculateFixedTokens();
+
+    if (fixed_tokens > model_context_size) {
+        throw std::runtime_error("Fixed tokens exceed model context size");
+    }
+
+    available_tokens = model_context_size - fixed_tokens;
+}
+
+int LlmMinOrMax::calculateFixedTokens() const {
+    int fixed_tokens = 0;
+    fixed_tokens += Tiktoken::GetNumTokens(user_prompt);
+    fixed_tokens += Tiktoken::GetNumTokens(llm_min_or_max_template);
+    return fixed_tokens;
+}
+
+nlohmann::json LlmMinOrMax::GetElement(const nlohmann::json &tuples) {
+    inja::Environment env;
+    nlohmann::json data;
+    data["tuples"] = tuples;
+    data["user_prompt"] = user_prompt;
+    auto prompt = env.render(llm_min_or_max_template, data);
+
+    nlohmann::json settings;
+    auto response = ModelManager::CallComplete(prompt, model, settings);
+    return response["selected"];
+}
+
+nlohmann::json LlmMinOrMax::LlmMinOrMaxCall(nlohmann::json &tuples) {
+
+    if (tuples.size() == 1) {
+        return tuples[0]["content"];
+    }
+
+    auto num_tuples = tuples.size();
+    std::vector<int> batch_size_list;
+    auto used_tokens = 0;
+    auto batch_size = 0;
+    for (int i = 0; i < num_tuples; i++) {
+        used_tokens += Tiktoken::GetNumTokens(tuples[i].dump());
+        batch_size++;
+        if (used_tokens >= available_tokens) {
+            batch_size_list.push_back(batch_size);
+            used_tokens = 0;
+            batch_size = 0;
+        } else if (i == num_tuples - 1) {
+            batch_size_list.push_back(batch_size);
+        }
+    }
+
+    auto responses = nlohmann::json::array();
+    for (int i = 0; i < batch_size_list.size(); i++) {
+        auto start_index = i * batch_size;
+        auto end_index = start_index + batch_size;
+        auto batch = nlohmann::json::array();
+        for (int j = start_index; j < end_index; j++) {
+            batch.push_back(tuples[j]);
+        }
+        auto ranked_indices = GetElement(batch);
+        responses.push_back(batch[ranked_indices.get<int>()]);
+    }
+    return LlmMinOrMaxCall(responses);
```

Comment:
This recursive implementation is much harder to read than necessary.
It would be easier to read:

```auto start_pos = 0;
     while (tuples not full processed [start_pos < tuples.size() ?]) {
         end_pos = find_end_pos_to_fit_within_context;
         GetElement taking tuples with start_pos and end_pos, you get max/min tuple
         Write the tuple into end_pos with a comment explaining you store result there.
         start_pos = end_pos
     }
```

Reply:
This code removes the need for responses and the need for push_back etc. It would always simplify the batch_size computation. You only need to find the next end based on the tuples you are about to process and not ALL tuples.